src/types/keyValue/keyValue.middleware.js

```diff
@@ -8,105 +8,77 @@ export const initState = {
   initialized: false,
 }
 
-const keyAlreadyExists =
-  state => (key, instanceKey) => state.array.find(o => o[key] === instanceKey)
+const keyAlreadyExists = state => key => state.keys.includes(key)
 
-const set = (key, payload) => ({
-  data: keyBy(payload, key),
-  keys: uniq(payload.map(element => element[key])),
-  array: payload,
+const mapDataToState = state => data => ({
+  ...state,
+  data,
+  keys: Object.keys(data),
+  array: Object.values(data),
   initialized: true,
 })
 
-const add = (key, state, payload) => {
-  let array
-  const instanceKey = payload[key]
-  if (!keyAlreadyExists(state)(key, instanceKey)) {
-    array = [...state.array, payload]
-  } else {
-    array = state.array.map(o => (o[key] === instanceKey ? payload : o))
-  }
+const set = (key, state, payload) => {
+  const data = [].concat(payload).reduce(
+    (acc, curr) => ({ ...acc, [curr[key]]: curr }),
+    {},
+  )
```

Comment:
I have done some benchmark for this aproach. I found that doing this rather than the a mutable aproach is from far less effective. It also a litle slower to use `concat` instead of checking if payload is an array.


**Mutable implementation**
```es6
let data = {}
[].concat(payload).forEach(entity => {
  data[entity[key]] = entity
})
```

**Mutable without concat**
```es6
let data = {}
(Array.isArray(test) : test : [test]).forEach(entity => {
  data[entity[key]] = entity
})
```

[You can find the benchmark here](https://jsperf.com/test-k-redux-factory-1/1)

Reply:
Yes ... I know... but this is not beautiful :(

Reply:
Yes I know... But in facts, I have found that the `reduce` solution is more dificult to read and understand for must of other developers.

I think we will not have a better solution since a native JS syntax is defined to do what we want...